### PR TITLE
[6004] show error when org/group type is changed

### LINKED ISSUE #6004: Can't change group type with organization/group update/patch operations
## CKAN version

CKAN 2.8, CKAN 2.9

## Describe the bug

I recently wanted to change the type of an organisation on a CKAN instance via the API. I made an API call like this:

```
curl -X POST http://my-ckan-server.com/api/3/action/organization_patch  -H "Authorization: <my-api-key>" -H "Content-Type: application/json" -H "Accept: application/json" --data '{"id": "<org-id>", "type": "new-type"}'
```

This returned a response:

```
200 OK

{
  "success": true,
  "result": {
    ...
    "type": "old-type",
  }
}
```

and did not update the type field.

This happens because the `type` field in the request body is ignored here:

https://github.com/ckan/ckan/blob/170556a97f38d42b94474943ccab4cef19fd9d49/ckan/logic/action/update.py#L645

It would be useful if it were possible to change the group type using the API for appropriately authorized users.
If it is impossible to do this by design, it would be much clearer what is going on if this request failed with a response like:

```
400 Bad Request

{
  "success": false,
  "error": {
    "message": "type can't be updated via the API",
  }
}
```

instead of reporting success but not making the change.

## Steps to reproduce

```
curl -X POST http://my-ckan-server.com/api/3/action/organization_patch  -H "Authorization: <my-api-key>" -H "Content-Type: application/json" -H "Accept: application/json" --data '{"id": "<org-id>", "type": "new-type"}'
```

## Expected behavior

Either the object should be updated, or a useful error response should be returned

### PR DESCRIPTION
**Pull request by robin-NEC**
Fixes #6004 

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

File: ckan/logic/action/update.py [hunk 1]
```diff
@@ -651,8 +651,13 @@ def _group_or_org_update(
     if group is None:
         raise NotFound('Group was not found.')
     context["group"] = group
-
-    data_dict['type'] = group.type
+    
+    data_dict_type = data_dict.get('type', None) 
+    if data_dict_type is None:
+        data_dict_type = group.type
```
Comment (pdelboca) [inline ckan/logic/action/update.py:657]: What's this assignment for? Where are we using this variable?

**Review by pdelboca — COMMENTED** [review]
Thanks for this PR @robin-NEC ! 

I left some comments. Basically if the main goal is to warn users of type updates it's enough to just compare what's coming in `data_dict` and raise a validation error if the type is different.

**Comment by robin-NEC** [comment]
> Thanks for this PR @robin-NEC ! I left some comments.
> 
> Basically if the main goal is to warn users of type updates it's enough to just compare what's coming in `data_dict` and raise a validation error if the type is different.

Hi @pdelboca , I believe what you're suggesting is this commit:  https://github.com/ckan/ckan/pull/6978/commits/1c68fa7326a79147d392203802e7c0bae3503895.

It seems that in test_organization.py some test cases fails.

``` 
    def test_all_fields_saved(self, app, initial_data):
>       app.post(
            url=url_for(
                "organization.edit", id=initial_data["organization"]["id"]
            ),
            extra_environ=initial_data["user_env"],
            data={
                "name": u"all-fields-edited",
                "title": "Science",
                "description": "Sciencey datasets",
                "image_url": "http://example.com/image.png",
                "save": ""
            }
        )

ckan/tests/controllers/test_organization.py:131: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ckan/tests/helpers.py:210: in post
    res = self.test_client().post(url, *args, **kwargs)
../../lib/python3.8/site-packages/werkzeug/test.py:1136: in post
    return self.open(*args, **kw)
ckan/tests/helpers.py:244: in open
    res = super(CKANTestClient, self).open(*args, **kwargs)
../../lib/python3.8/site-packages/werkzeug/test.py:1076: in open
    response = self.run_wsgi_app(request.environ, buffered=buffered)
../../lib/python3.8/site-packages/werkzeug/test.py:945: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
../../lib/python3.8/site-packages/werkzeug/test.py:1233: in run_wsgi_app
    app_rv = app(environ, start_response)
ckan/config/middleware/common_middleware.py:68: in __call__
    return self.app(environ, start_response)
ckan/config/middleware/flask_app.py:69: in __call__
    return self.app(environ, start_response)
ckan/config/middleware/common_middleware.py:87: in __call__
    return self.app(environ, start_response)
../../lib/python3.8/site-packages/repoze/who/middleware.py:64: in __call__
    return self.app(environ, start_response)
ckan/config/middleware/flask_app.py:92: in __call__
    response = request.get_response(self.app)
../../lib/python3.8/site-packages/webob/request.py:1313: in send
    status, headers, app_iter = self.call_application(
../../lib/python3.8/site-packages/webob/request.py:1278: in call_application
    app_iter = application(self.environ, start_response)
../../lib/python3.8/site-packages/flask/app.py:2091: in __call__
    return self.wsgi_app(environ, start_response)
../../lib/python3.8/site-packages/beaker/middleware.py:156: in __call__
    return self.wrap_app(environ, session_start_response)
ckan/config/middleware/common_middleware.py:31: in __call__
    return self.app(environ, start_response)
../../lib/python3.8/site-packages/flask/app.py:2076: in wsgi_app
    response = self.handle_exception(e)
../../lib/python3.8/site-packages/flask/app.py:2073: in wsgi_app
    response = self.full_dispatch_request()
../../lib/python3.8/site-packages/flask/app.py:1518: in full_dispatch_request
    rv = self.handle_user_exception(e)
../../lib/python3.8/site-packages/flask/app.py:1516: in full_dispatch_request
    rv = self.dispatch_request()
../../lib/python3.8/site-packages/flask/app.py:1502: in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
../../lib/python3.8/site-packages/flask/views.py:84: in view
    return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
../../lib/python3.8/site-packages/flask/views.py:158: in dispatch_request
    return current_app.ensure_sync(meth)(*args, **kwargs)
ckan/views/group.py:929: in post
    group = _action(u'group_update')(context, data_dict)
ckan/logic/__init__.py:527: in wrapped
    result = _action(context, data_dict, **kw)
ckan/logic/action/update.py:785: in organization_update
    return _group_or_org_update(context, data_dict, is_org=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

context = {'__auth_audit': [('organization_update', 140143678672608)], '__auth_user_obj_checked': True, 'allow_partial_update': ...il_notifications=False sysadmin=False state=active image_url=https://placeimg.com/190/864/any plugin_extras=None>, ...}
data_dict = {'description': 'Sciencey datasets', 'id': '9433d79f-f519-4c62-9104-1350ffb23423', 'image_url': 'http://example.com/image.png', 'name': 'all-fields-edited', ...}
is_org = True

    def _group_or_org_update(
            context: Context, data_dict: DataDict, is_org: bool = False):
        model = context['model']
        session = context['session']
        id = _get_or_bust(data_dict, 'id')
    
        group = model.Group.get(id)
        if group is None:
            raise NotFound('Group was not found.')
        context["group"] = group
    
>       if data_dict['type'] != group.type:
E       KeyError: 'type'

ckan/logic/action/update.py:666: KeyError
------------------------------------------- Captured stdout setup -------------------------------------------
2022-08-29 13:10:45,161 CRITI [ckan.lib.uploader] Please specify a ckan.storage_path in your config
                        for your uploads
2022-08-29 13:10:45,303 CRITI [ckan.lib.uploader] Please specify a ckan.storage_path in your config
                        for your uploads
2022-08-29 13:10:45,382 CRITI [ckan.lib.uploader] Please specify a ckan.storage_path in your config
                        for your uploads
-------------------------------------------- Captured log setup ---------------------------------------------
CRITICAL ckan.lib.uploader:uploader.py:85 Please specify a ckan.storage_path in your config
                        for your uploads
CRITICAL ckan.lib.uploader:uploader.py:85 Please specify a ckan.storage_path in your config
                        for your uploads
CRITICAL ckan.lib.uploader:uploader.py:85 Please specify a ckan.storage_path in your config
                        for your uploads
========================================== short test summary info ==========================================
FAILED ckan/tests/controllers/test_organization.py::TestOrganizationEdit::test_saved - KeyError: 'type'
FAILED ckan/tests/controllers/test_organization.py::TestOrganizationEdit::test_all_fields_saved - KeyError...
======================================= 2 failed, 28 passed in 55.76s =======================================

```
It seems this line :
https://github.com/ckan/ckan/blob/170556a97f38d42b94474943ccab4cef19fd9d49/ckan/logic/action/update.py#L645
not only overrides the value `type` added in `data_dict` but also sets it if it is not present in the `data_dict` and in absence of it, the test case fails.

**Comment by robin-NEC** [comment]
hi @pdelboca , the below error 
```
_____________________ TestSearch.test_search_page_results ______________________

self = <ckan.tests.controllers.test_package.TestSearch object at 0x7f688a338210>
app = <ckan.tests.helpers.CKANTestApp object at 0x7f68835f2610>

    def test_search_page_results(self, app):
        """Searching for datasets returns expected results."""
    
        factories.Dataset(name="dataset-one", title="Dataset One")
        factories.Dataset(name="dataset-two", title="Dataset Two")
        factories.Dataset(name="dataset-three", title="Dataset Three")
    
        search_url = url_for("dataset.search")
        search_results = app.get(search_url, query_string={'q': 'One'})
    
>       assert "1 dataset found" in search_results
E       AssertionError: assert '1 dataset found' in <WrapperTestResponse 16437 bytes [200 OK]>

ckan/tests/controllers/test_package.py:1538: AssertionError
```
is not associated with my fix.

**Pull request merged**